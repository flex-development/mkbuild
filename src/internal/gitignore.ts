/**
 * @file Internal - gitignore
 * @module mkbuild/internal/gitignore
 */

import * as mlly from '@flex-development/mlly'
import * as pathe from '@flex-development/pathe'

/**
 * Returns a set of ignore patterns found in a `.gitignore` file.
 *
 * @async
 *
 * @param {string} absWorkingDir - Absolute path to current working directory
 * @return {Promise<Set<string>>} Ignore patterns set
 */
const gitignore = async (absWorkingDir: string): Promise<Set<string>> => {
  /**
   * Ignore patterns found in `.gitignore` file.
   *
   * @const {Set<string>} ignore
   */
  const ignore: Set<string> = new Set<string>()

  // try adding ignore patterns from .gitignore
  try {
    /**
     * Absolute path to `.gitignore` file.
     *
     * @const {string} path
     */
    const path: string = pathe.resolve(absWorkingDir, '.gitignore')

    /**
     * `.gitignore` file content.
     *
     * @const {string} gitignore
     */
    const content: string = (await mlly.getSource(path)) as string

    // add ignore patterns from .gitignore
    for (const line of content.split(/\r?\n/)) {
      if (!line.trim()) continue
      if (line.startsWith('#') || line.startsWith('!')) continue
      ignore.add(line.trim())
    }
  } catch {
    // do nothing if .gitignore file was not found
  }

  return ignore
}

export default gitignore
