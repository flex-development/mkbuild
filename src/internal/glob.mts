/**
 * @file Internal - glob
 * @module mkbuild/internal/glob
 */

import dfs from '#internal/fs'
import toPath from '#internal/to-path'
import type { DirectoryContent, File } from '@flex-development/fst'
import {
  fromFileSystem,
  type Dirent,
  type Options
} from '@flex-development/fst-util-from-fs'
import pathe from '@flex-development/pathe'
import type { Type } from '@flex-development/unist-util-types'
import { ok } from 'devlop'

export default glob

/**
 * Glob options.
 *
 * @internal
 */
interface GlobOptions extends Options {
  /**
   * Path filters to determine if nodes should be added to the tree.
   *
   * @override
   */
  filters?: null | undefined

  /**
   * Glob pattern, or a list of patterns, used to exclude matches.
   */
  ignore?: Set<string> | readonly string[] | string | null | undefined

  /**
   * Glob pattern, or a list of patterns, matching files to include in the tree.
   *
   * > ðŸ‘‰ **Note**: Patterns matching directories to search will be generated
   * > using this list.
   */
  include?: Set<string> | readonly string[] | string | null | undefined
}

/**
 * Traverse the file system and return a list of pathnames matching a a defined
 * set of patterns.
 *
 * @internal
 *
 * @this {void}
 *
 * @param {GlobOptions | null | undefined} [options]
 *  Glob options
 * @return {string[]}
 *  List of matched files
 */
function glob(this: void, options?: GlobOptions | null | undefined): string[] {
  options ??= {}

  if (typeof options.ignore === 'string') options.ignore = [options.ignore]
  if (typeof options.include === 'string') options.include = [options.include]

  /**
   * List of matched files.
   *
   * @const {string[]} files
   */
  const files: string[] = []

  /**
   * List of glob patterns used to exclude matches.
   *
   * @const {Set<string>} ignore
   */
  const ignore: Set<string> = new Set<string>(options.ignore)

  /**
   * Record of glob patterns matching directories to search and files to
   * include.
   *
   * @const {Record<Type<DirectoryContent>, Set<string>>} matchers
   */
  const matchers: Record<Type<DirectoryContent>, Set<string>> = {
    directory: new Set(),
    file: new Set(options.include)
  }

  // convert negated patterns to ignore patterns
  for (const pattern of matchers.file) {
    if (pattern.startsWith('!')) {
      ignore.add(pattern.slice(1))
      matchers.file.delete(pattern)
    }
  }

  // get patterns matching directories to search
  for (const pattern of matchers.file) {
    if (pattern === '**') {
      matchers.directory.add(pattern)
    } else {
      /**
       * Directory name of {@linkcode pattern}.
       *
       * @const {string} dirname
       */
      let dirname: string = pathe.dirname(pattern)

      while (dirname !== pathe.dot) {
        matchers.directory.add(dirname)
        dirname = pathe.dirname(dirname)
      }
    }
  }

  // create matched file tree
  fromFileSystem({
    ...options,
    filters: {
      /**
       * Determine if a `directory` node should be added to the tree.
       *
       * @this {void}
       *
       * @param {string} x
       *  Relative path to directory
       * @return {boolean}
       *  `true` if `directory` node should be added, `false` otherwise
       */
      directory(this: void, x: string): boolean {
        return filter(x, matchers.directory)
      },
      /**
       * Determine if a `file` node should be added to the tree.
       *
       * @this {void}
       *
       * @param {string} x
       *  Relative path to file
       * @return {boolean}
       *  `true` if `file` node should be added, `false` otherwise
       */
      file(this: void, x: string): boolean {
        return filter(x, matchers.file)
      }
    },
    fs: { ...dfs, ...options.fs },
    handles: {
      /**
       * Add matched files.
       *
       * @this {void}
       *
       * @param {File} node
       *  File node
       * @param {Dirent} dirent
       *  Dirent object representing the file `node` was created from
       * @return {undefined}
       */
      file(this: void, node: File, dirent: Dirent): undefined {
        return void files.push(pathe.join(dirent.parentPath, node.name))
      }
    }
  })

  return files

  /**
   * @this {void}
   *
   * @param {string} x
   *  The path to match
   * @param {Set<string>} patterns
   *  List of glob patterns
   * @return {boolean}
   *  `true` if `x` matches include pattern and is not excluded
   */
  function filter(this: void, x: string, patterns: Set<string>): boolean {
    ok(options, 'expected `options`')

    return pathe.matchesGlob(x, [...patterns], {
      cwd: options.root ? toPath(options.root) : pathe.cwd(),
      dot: true,
      ignore: [...ignore]
    })
  }
}
