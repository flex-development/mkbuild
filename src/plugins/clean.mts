/**
 * @file Plugins - clean
 * @module mkbuild/plugins/clean
 */

import pathe from '@flex-development/pathe'
import { ok } from 'devlop'
import type {
  NormalizedOutputOptions,
  Plugin,
  PluginContext
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:clean'

/**
 * Cleaned directory cache.
 *
 * @internal
 *
 * @const {Set<string>} directories
 */
const directories: Set<string> = new Set<string>()

plugin.PLUGIN_NAME = PLUGIN_NAME
plugin.directories = directories

export default plugin

/**
 * Create an output directory cleaner plugin.
 *
 * @internal
 *
 * @see {@linkcode Plugin}
 *
 * @this {void}
 *
 * @return {Plugin}
 *  Output directory cleaner plugin
 */
function plugin(this: void): Plugin {
  return {
    name: PLUGIN_NAME,
    renderStart: { handler: renderStart, order: 'pre', sequential: true }
  }

  /**
   * Remove and recreate an output directory.
   *
   * @see https://rollupjs.org/plugin-development/#renderstart
   *
   * @async
   *
   * @this {PluginContext}
   *
   * @param {NormalizedOutputOptions} output
   *  Normalized output options
   * @return {Promise<undefined>}
   */
  async function renderStart(
    this: PluginContext,
    output: NormalizedOutputOptions
  ): Promise<undefined> {
    ok(typeof output.clean === 'boolean', 'expected `output.clean`')
    ok(typeof output.fs === 'object', 'expected `output.fs`')
    ok(typeof output.root === 'string', 'expected `output.root`')
    ok(typeof output.write === 'boolean', 'expected `output.write`')

    if (
      output.clean &&
      output.root &&
      output.write &&
      typeof output.dir === 'string'
    ) {
      directories.add(output.root)

      /**
       * Absolute path to output directory.
       *
       * @const {string} path
       */
      const path: string = pathe.join(output.root, output.dir)

      if (!directories.has(path)) {
        await output.fs.rm(path, { force: true, recursive: true })
        await output.fs.mkdir(path, { recursive: true })
      }

      directories.add(path)
    }

    return void this
  }
}
