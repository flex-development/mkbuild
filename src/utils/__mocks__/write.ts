/**
 * @file Mock Utilities - write
 * @module mkbuild/utils/mocks/write
 */

/**
 * `write` module type.
 */
type Actual = typeof import('../write')

/**
 * `write` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('../write')

export default vi.fn(actual.default)
