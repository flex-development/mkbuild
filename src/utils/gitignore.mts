/**
 * @file Utilities - gitignore
 * @module utils/gitignore
 */

import dfs from '#internal/fs'
import toPath from '#internal/to-path'
import type { FileSystem } from '@flex-development/mkbuild'
import { isFile } from '@flex-development/mlly'
import pathe from '@flex-development/pathe'
import { ok } from 'devlop'

/**
 * Extract a list of ignore patterns from a `.gitignore` file.
 *
 * @see {@linkcode FileSystem}
 *
 * @category
 *  utils
 *
 * @this {void}
 *
 * @param {URL | string | null | undefined} [dir]
 *  Module id of directory containing `.gitignore` file
 * @param {Partial<FileSystem> | null | undefined} [fs]
 *  File system adapter
 * @return {Set<string>}
 *  List of ignore patterns
 */
function gitignore(
  this: void,
  dir?: URL | string | null | undefined,
  fs?: Partial<FileSystem> | null | undefined
): Set<string> {
  fs = { ...dfs, ...fs }

  ok(typeof fs.readFileSync === 'function', 'expected `fs.readFileSync`')
  ok(typeof fs.statSync === 'function', 'expected `fs.statSync`')

  /**
   * Ignore patterns found in `.gitignore` file.
   *
   * @const {Set<string>} ignore
   */
  const ignore: Set<string> = new Set<string>()

  /**
   * Absolute path to `.gitignore` file.
   *
   * @const {string} path
   */
  const path: string = pathe.join(toPath(dir ?? pathe.cwd()), '.gitignore')

  if (isFile(path, fs as FileSystem)) {
    for (let line of fs.readFileSync(path, 'utf8').split(/\r?\n/)) {
      line = line.trim()

      if (!line || line.startsWith('#') || line.startsWith('!')) continue
      if (line.endsWith(pathe.sep)) line += '*'.repeat(2)
      ignore.add(line)
    }
  }

  return ignore
}

export default gitignore
