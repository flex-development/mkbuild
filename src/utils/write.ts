/**
 * @file Utilities - write
 * @module mkbuild/utils/write
 */

import type { Config } from '#src/interfaces'
import type { OutputFile } from 'esbuild'
import fse from 'fs-extra'
import * as pathe from 'pathe'

/**
 * Writes a build result to the file system.
 *
 * @template R - Build result object type
 *
 * @async
 *
 * @param {R} result - Build result object
 * @param {Config['fs']} [fs=fse] - Custom file system methods
 * @return {Promise<R>} Written build result
 */
async function write<R extends OutputFile = OutputFile>(
  result: R,
  fs: Config['fs'] = fse
): Promise<R> {
  // create subdirectories in outdir
  await fs.mkdirp(pathe.dirname(result.path))

  // write build result
  await fs.writeFile(result.path, result.text, 'utf8')

  return result
}

export default write
