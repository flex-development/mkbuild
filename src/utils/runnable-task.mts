/**
 * @file Utilities - runnableTask
 * @module mkbuild/utils/runnableTask
 */

import createEntryFileNames from '#internal/create-entry-file-names'
import createOnLog from '#internal/create-on-log'
import formatLog from '#internal/format-log'
import dfs from '#internal/fs'
import glob from '#internal/glob'
import loadTsconfig from '#internal/load-tsconfig'
import readPackageJson from '#internal/read-package-json'
import toPath from '#internal/to-path'
import withTrailingSlash from '#internal/with-trailing-slash'
import clean from '#plugins/clean'
import dts from '#plugins/dts'
import esbuild from '#plugins/esbuild'
import metadata from '#plugins/metadata'
import resolve from '#plugins/resolve'
import write from '#plugins/write'
import type WithUndefined from '#types/with-undefined'
import formatToExt from '#utils/format-to-ext'
import gitignore from '#utils/gitignore'
import IGNORE_PATTERNS from '#utils/ignore-patterns'
import TASK_PLUGIN_NAME from '#utils/task-plugin-name'
import type {
  FileSystem,
  Input,
  Message,
  OutputAsset,
  OutputChunk,
  Result,
  RunnableTask,
  Task
} from '@flex-development/mkbuild'
import pathe from '@flex-development/pathe'
import type { PackageJson } from '@flex-development/pkg-types'
import {
  ifelse,
  isNIL,
  ksort
} from '@flex-development/tutils'
import { ok } from 'devlop'
import {
  rollup,
  type InputOptions,
  type InputPluginOption,
  type MinimalPluginContext,
  type NormalizedInputOptions,
  type NormalizedOutputOptions,
  type OutputOptions,
  type PluginContext,
  type RollupBuild
} from 'rollup'
import { createMerger } from 'smob'

/**
 * Create a runnable build task.
 *
 * @see {@linkcode FileSystem}
 * @see {@linkcode RunnableTask}
 * @see {@linkcode Task}
 *
 * @todo bundle mode support
 *
 * @category
 *  utils
 *
 * @this {void}
 *
 * @param {Task | null | undefined} [task]
 *  The build task to handle
 * @param {Partial<FileSystem> | null | undefined} [fs]
 *  File system API
 * @return {RunnableTask}
 *  Runnable build task
 */
function runnableTask(
  this: void,
  task?: Task | null | undefined,
  fs?: Partial<FileSystem> | null | undefined
): RunnableTask {
  task ??= {}
  fs = { ...dfs, ...fs }

  /**
   * Entry points.
   *
   * @var {Input | undefined} input
   */
  let input: Input | undefined

  /**
   * Build messages.
   *
   * @var {Message[]} messages
   */
  let messages: Message[]

  /**
   * Package manifest.
   *
   * @var {PackageJson} pkg
   */
  let pkg: PackageJson

  /**
   * Input plugins.
   *
   * @var {InputPluginOption[]} plugins
   */
  let plugins: InputPluginOption[]

  return ksort(Object.assign({}, task, { run }) as RunnableTask)

  /**
   * Attach additional data to `options`.
   *
   * @see https://rollupjs.org/plugin-development/#buildstart
   *
   * @async
   *
   * @this {PluginContext}
   *
   * @param {NormalizedInputOptions} options
   *  Normalized input options
   * @return {Promise<undefined>}
   */
  async function buildStart(
    this: PluginContext,
    options: NormalizedInputOptions
  ): Promise<undefined> {
    ok(fs, 'expected `fs`')
    ok(task, 'expected `task`')
    ok(typeof task.root === 'string', 'expected `task.root`')

    options.fs = fs as FileSystem
    options.pkg = pkg
    options.root = task.root

    options.tsconfig = createMerger({ priority: 'right' })(await loadTsconfig({
      cwd: task.root,
      fs: options.fs,
      tsconfig: task.tsconfig,
      tsconfigRaw: task.tsconfigRaw
    }), {
      compilerOptions: {
        noEmit: false,
        noEmitOnError: false,
        noErrorTruncation: true
      }
    })

    return void this
  }

  /**
   * Set input options.
   *
   * @see https://rollupjs.org/plugin-development/#options
   *
   * @this {MinimalPluginContext}
   *
   * @param {{ plugins?: InputPluginOption }} options
   *  Options passed to `rollup`
   * @return {WithUndefined<InputOptions>}
   *  Rollup input options
   */
  function options(
    this: MinimalPluginContext,
    options: { plugins?: InputPluginOption }
  ): WithUndefined<InputOptions> {
    ok(task, 'expected `task`')
    ok(typeof task.format === 'string', 'expected `task.format`')
    ok(typeof task.logLevel === 'string', 'expected `task.logLevel`')
    ok(typeof task.root === 'string', 'expected `task.root`')
    ok(Array.isArray(options.plugins), 'expected `options.plugins`')

    plugins = []

    task.dts ??= Object.keys({ ...pkg.devDependencies }).includes('typescript')
    task.gitignore ??= true
    task.ignore ??= new Set(IGNORE_PATTERNS)
    task.sourcemapExcludeSources ??= true
    task.strictDeprecations ??= true

    if (Array.isArray(input) || typeof input === 'string') {
      input = glob({
        fs,
        ignore: new Set([
          ...task.gitignore ? gitignore(task.root, fs) : [],
          ...task.ignore
        ]),
        include: input,
        root: task.root
      })
    }

    plugins.push(
      ...options.plugins,
      task.plugins,
      resolve({ ...task.resolve }),
      esbuild({
        ...task.esbuild,
        format: task.format,
        sourceRoot: task.sourcemapBaseUrl,
        sourcesContent: !task.sourcemapExcludeSources
      }),
      task.dts && dts({ only: task.dts === 'only', path: task.typescriptPath }),
      metadata(),
      clean(),
      write()
    )

    return {
      cache: false,
      input,
      jsx: false,
      logLevel: ifelse(task.logLevel === 'silent', undefined, task.logLevel),
      makeAbsoluteExternalsRelative: undefined,
      maxParallelFileOps: task.maxParallelFileOps ?? 20,
      onLog: createOnLog(messages),
      perf: task.experimental?.perf ?? undefined,
      plugins,
      preserveSymlinks: task.resolve?.preserveSymlinks ?? false,
      strictDeprecations: task.strictDeprecations,
      watch: false
    }
  }

  /**
   * Compare the filenames of `a` and `b`.
   *
   * @param {OutputAsset | OutputChunk} a
   *  First output asset or chunk
   * @param {OutputAsset | OutputChunk} b
   *  Second output asset or chunk
   * @return {number}
   *  Comparison result
   */
  function outputCompare(
    a: OutputAsset | OutputChunk,
    b: OutputAsset | OutputChunk
  ): number {
    return a.fileName.localeCompare(b.fileName)
  }

  /**
   * Set output options.
   *
   * @see https://rollupjs.org/plugin-development/#outputoptions
   *
   * @this {MinimalPluginContext}
   *
   * @return {WithUndefined<OutputOptions>}
   *  Rollup output options
   */
  function outputOptions(
    this: MinimalPluginContext
  ): WithUndefined<OutputOptions> {
    ok(task, 'expected `task`')
    ok(typeof task.format === 'string', 'expected `task.format`')
    ok(typeof task.outdir === 'string', 'expected `task.outdir`')

    task.assetFileNames ??= 'assets/[name][extname]'
    task.clean ??= true
    task.entryFileNames ??= '[name]{extname}'
    task.sanitizeFileName ??= true
    task.write ??= false

    if (!task.ext) task.ext = formatToExt(task.format)
    task.ext = pathe.formatExt(task.ext)

    return {
      assetFileNames: task.assetFileNames,
      banner: task.banner ?? undefined,
      dir: task.outdir,
      entryFileNames: createEntryFileNames(task),
      footer: task.footer ?? undefined,
      format: task.format,
      hashCharacters: task.hashCharacters ?? 'base64',
      preserveModules: true,
      sanitizeFileName: task.sanitizeFileName,
      sourcemap: task.sourcemap ?? undefined,
      sourcemapBaseUrl: ifelse(
        isNIL(task.sourcemapBaseUrl),
        undefined,
        String(task.sourcemapBaseUrl)
      ),
      sourcemapExcludeSources: task.sourcemapExcludeSources!,
      sourcemapIgnoreList: task.sourcemapIgnoreList ?? undefined,
      sourcemapPathTransform: task.sourcemapPathTransform ?? undefined,
      validate: task.validate ?? undefined,
      virtualDirname: task.virtualDirname ?? undefined
    }
  }

  /**
   * Attach additional data to `output` options.
   *
   * @see https://rollupjs.org/plugin-development/#renderstart
   *
   * @this {PluginContext}
   *
   * @param {NormalizedOutputOptions} output
   *  Normalized output options
   * @return {undefined}
   */
  function renderStart(
    this: PluginContext,
    output: NormalizedOutputOptions
  ): undefined {
    ok(fs, 'expected `fs`')
    ok(task, 'expected `task`')
    ok(typeof task.clean === 'boolean', 'expected `task.clean`')
    ok(typeof task.root === 'string', 'expected `task.root`')
    ok(typeof task.write === 'boolean', 'expected `task.write`')

    output.clean = task.clean
    output.fs = fs as FileSystem
    output.root = task.root
    output.write = task.write

    return void this
  }

  /**
   * Run {@linkcode task}.
   *
   * @async
   *
   * @return {Promise<Result>}
   *  Build result
   */
  async function run(): Promise<Result> {
    ok(fs, 'expected `fs`')
    ok(task, 'expected `task`')

    task.root = toPath(task.root ?? pathe.cwd())
    task.root = withTrailingSlash(pathe.resolve(task.root))

    input = task.input ?? undefined
    messages = []
    pkg = readPackageJson(task.root, fs as FileSystem)

    task.format ??= pkg.type === 'module' ? 'esm' : 'cjs'
    task.logLevel ??= 'info'
    task.outdir ??= 'dist'

    /**
     * Build result.
     *
     * @const {Result} result
     */
    const result: Result = Object.defineProperties({
      failure: null,
      format: task.format,
      messages,
      outdir: task.outdir,
      outputs: [],
      pkg,
      root: task.root,
      size: 0,
      task: task as Result['task'],
      timings: null
    }, {
      pkg: {
        enumerable: false,
        value: pkg,
        writable: true
      },
      size: {
        enumerable: true,

        /**
         * Get the size of `this` build result.
         *
         * @this {Result}
         *
         * @return {number}
         *  Build result size in bytes
         */
        get(this: Result): number {
          return this.outputs.reduce((acc, output) => {
            ok(typeof output.bytes === 'number', 'expected `output.bytes`')
            return acc + output.bytes
          }, 0)
        }
      }
    })

    try {
      /**
       * Rollup build.
       *
       * @const {RollupBuild} build
       */
      const build: RollupBuild = await rollup({
        plugins: [
          {
            api: { fs, pkg },
            // @ts-expect-error [2353] vite-only option
            apply: 'build',
            buildStart: { handler: buildStart, order: 'pre', sequential: true },
            name: TASK_PLUGIN_NAME,
            onLog: { handler: formatLog, order: 'pre' },
            options: { handler: options, order: 'pre', sequential: true },
            outputOptions: { handler: outputOptions, order: 'pre' },
            renderStart: { handler: renderStart, order: 'pre' }
          }
        ]
      })

      // generate rollup build outputs and add outputs to build result
      result.outputs = (await build.generate({})).output.sort(outputCompare)

      // close rollup build so `closeBundle` hooks are called
      void await build.close()

      // add performance timings to build result
      if (build.getTimings) result.timings = build.getTimings()
    } catch (e: unknown) {
      result.failure = e as Error
    }

    // rollup will throw if `task.input` is empty, `null`, or `undefined`
    if (result.failure?.message.match(/must supply options.input to rollup/)) {
      result.failure = null
    }

    ok(Array.isArray(plugins), 'expected `plugins`')

    result.task.input = input
    result.task.plugins = plugins

    return ksort(result.task), result
  }
}

export default runnableTask
