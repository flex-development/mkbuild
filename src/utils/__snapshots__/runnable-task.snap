// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/runnableTask > #run > transpile > should return build result (4) 1`] = `
[
  {
    "bytes": 52,
    "code": "export { default as reverse } from './reverse.mjs';
",
    "dynamicImports": [],
    "exports": [
      "reverse",
    ],
    "facadeModuleId": \${process.cwd()}/__fixtures__/pkg/reverse/src/index.mts,
    "fileName": "index.mjs",
    "implicitlyLoadedBefore": [],
    "importedBindings": {
      "reverse.mjs": [
        "default",
      ],
    },
    "imports": [
      "reverse.mjs",
    ],
    "isDynamicEntry": false,
    "isEntry": true,
    "isImplicitEntry": false,
    "map": null,
    "moduleIds": [
      \${process.cwd()}/__fixtures__/pkg/reverse/src/index.mts,
    ],
    "modules": {
      \${process.cwd()}/__fixtures__/pkg/reverse/src/index.mts: {
        "code": null,
        "originalLength": 131,
        "removedExports": [],
        "renderedExports": [
          "reverse",
        ],
        "renderedLength": 0,
      },
    },
    "name": "index",
    "preliminaryFileName": "index.mjs",
    "referencedFiles": [],
    "sourcemapFileName": null,
    "type": "chunk",
  },
  {
    "bytes": 79,
    "code": "var max_value_default = 2 ** 31 - 1;

export { max_value_default as default };
",
    "dynamicImports": [],
    "exports": [
      "default",
    ],
    "facadeModuleId": \${process.cwd()}/__fixtures__/pkg/reverse/src/max-value.mts,
    "fileName": "max-value.mjs",
    "implicitlyLoadedBefore": [],
    "importedBindings": {},
    "imports": [],
    "isDynamicEntry": false,
    "isEntry": true,
    "isImplicitEntry": false,
    "map": null,
    "moduleIds": [
      \${process.cwd()}/__fixtures__/pkg/reverse/src/max-value.mts,
    ],
    "modules": {
      \${process.cwd()}/__fixtures__/pkg/reverse/src/max-value.mts: {
        "code": "var max_value_default = 2 ** 31 - 1;",
        "originalLength": 97,
        "removedExports": [],
        "renderedExports": [
          "default",
        ],
        "renderedLength": 36,
      },
    },
    "name": "max-value",
    "preliminaryFileName": "max-value.mjs",
    "referencedFiles": [],
    "sourcemapFileName": null,
    "type": "chunk",
  },
  {
    "bytes": 78,
    "code": "var min_value_default = (-2) ** 31;

export { min_value_default as default };
",
    "dynamicImports": [],
    "exports": [
      "default",
    ],
    "facadeModuleId": \${process.cwd()}/__fixtures__/pkg/reverse/src/min-value.mts,
    "fileName": "min-value.mjs",
    "implicitlyLoadedBefore": [],
    "importedBindings": {},
    "imports": [],
    "isDynamicEntry": false,
    "isEntry": true,
    "isImplicitEntry": false,
    "map": null,
    "moduleIds": [
      \${process.cwd()}/__fixtures__/pkg/reverse/src/min-value.mts,
    ],
    "modules": {
      \${process.cwd()}/__fixtures__/pkg/reverse/src/min-value.mts: {
        "code": "var min_value_default = (-2) ** 31;",
        "originalLength": 96,
        "removedExports": [],
        "renderedExports": [
          "default",
        ],
        "renderedLength": 35,
      },
    },
    "name": "min-value",
    "preliminaryFileName": "min-value.mjs",
    "referencedFiles": [],
    "sourcemapFileName": null,
    "type": "chunk",
  },
  {
    "bytes": 432,
    "code": "import max_value_default from './max-value.mjs';
import min_value_default from './min-value.mjs';

function reverse(x) {
  if (x >= -9 && x <= 9) return x;
  const digits = [...x.toString(), x < 0 ? "-" : ""];
  const reversed = Number.parseInt(digits.reverse().join(""));
  return reversed < min_value_default || reversed > max_value_default ? 0 : reversed;
}
var reverse_default = reverse;

export { reverse_default as default };
",
    "dynamicImports": [],
    "exports": [
      "default",
    ],
    "facadeModuleId": \${process.cwd()}/__fixtures__/pkg/reverse/src/reverse.mts,
    "fileName": "reverse.mjs",
    "implicitlyLoadedBefore": [],
    "importedBindings": {
      "max-value.mjs": [
        "default",
      ],
      "min-value.mjs": [
        "default",
      ],
    },
    "imports": [
      "max-value.mjs",
      "min-value.mjs",
    ],
    "isDynamicEntry": false,
    "isEntry": true,
    "isImplicitEntry": false,
    "map": null,
    "moduleIds": [
      \${process.cwd()}/__fixtures__/pkg/reverse/src/reverse.mts,
    ],
    "modules": {
      \${process.cwd()}/__fixtures__/pkg/reverse/src/reverse.mts: {
        "code": "function reverse(x) {
  if (x >= -9 && x <= 9) return x;
  const digits = [...x.toString(), x < 0 ? "-" : ""];
  const reversed = Number.parseInt(digits.reverse().join(""));
  return reversed < min_value_default || reversed > max_value_default ? 0 : reversed;
}
var reverse_default = reverse;",
        "originalLength": 1200,
        "removedExports": [],
        "renderedExports": [
          "default",
        ],
        "renderedLength": 292,
      },
    },
    "name": "reverse",
    "preliminaryFileName": "reverse.mjs",
    "referencedFiles": [],
    "sourcemapFileName": null,
    "type": "chunk",
  },
]
`;
